strawberryfields/backends/bosonicbackend/backend.py

```diff
@@ -54,6 +69,207 @@ def __init__(self):
+                    val = cmd.op.apply(cmd.reg, self, **kwargs)
+                    if val is not None:
+                        for i, r in enumerate(cmd.reg):
+                            samples_dict[r.ind] = val[:, i]
+
+                            # Internally also store all the measurement outcomes
+                            if r.ind not in all_samples:
+                                all_samples[r.ind] = list()
+                            all_samples[r.ind].append(val[:, i])
+
+                    applied.append(cmd)
+
+                except NotApplicableError:
+                    # command is not applicable to the current backend type
+                    raise NotApplicableError(
+                        "The operation {} cannot be used with the Bosonic Backend.".format(cmd.op)
+                    ) from None
+
+                except NotImplementedError:
+                    # command not directly supported by backend API
+                    raise NotImplementedError(
+                        "The operation {} has not been implemented in the Bosonic Backend for the arguments {}.".format(
+                            cmd.op, kwargs
+                        )
+                    ) from None
+
+        return applied, samples_dict, all_samples
+
+    def init_circuit(self, prog):
+        """Instantiate the circuit and initialize weights, means, and covs
+        depending on the Preparation classes.
+
+        Args:
+            prog (object): sf.Program instance
+
+        Raises:
+            NotImplementedError: if Ket or DensityMatrix preparation used
+        """
+        from strawberryfields.ops import (
+            Bosonic,
+            Catstate,
+            DensityMatrix,
+            Fock,
+            GKP,
+            Ket,
+            _New_modes,
+        )
+
+        nongausspreps = (Bosonic, Catstate, DensityMatrix, Fock, GKP, Ket)
+        nmodes = prog.init_num_subsystems
+        self.begin_circuit(nmodes)
+        # Dummy initial weights, means and covs
+        init_weights, init_means, init_covs = [[0] * nmodes for i in range(3)]
+
+        vac_means = np.zeros((1, 2), dtype=complex)  # .tolist()
+        vac_covs = np.array([0.5 * self.circuit.hbar * np.identity(2)])
+
+        # List of modes that have been traversed through
+        reg_list = []
+
+        # Go through the operations in the circuit
+        for cmd in prog.circuit:
+            # Check if an operation other than New() has already acted on these modes.
+            labels = [label.ind for label in cmd.reg]
+            isitnew = 1 - np.isin(labels, reg_list)
+            if np.any(isitnew):
+                # Operation parameters
+                pars = cmd.op.p
+                for reg in labels:
+                    # All the possible preparations should go in this loop
+                    if isinstance(cmd.op, Bosonic):
+                        weights, means, covs = [pars[i].tolist() for i in range(3)]
```

Comment:
Would it be possible to transform `pars[i]` to `ndarrays` (does `[np.array(pars[i]) for i in range(3)]` work instead)? It seems like most `weights`, `means` and `covs` below will be `ndarrays`, so just to be consistent and avoid potential issues down the road.

Reply:
Done.